README.md

```diff
@@ -1 +1,142 @@
+This lesson will cover the ...
+
+## Objectives
+
+1. Illustrate the basic async code (non-blocking I/O) vs. sync code, e.g., `setTimeout()` in Node
+2. Describe `process.nextTick()`
+3. Describe `setImmediate()`
+
+## Node Asynchronous Code
+
+One of the major selling point (a.k.a. benefits) of Node is it's non-blocking I/O which is powered by asynchronous code. Writing asynchronous code in JavaScript is straightforward, but for developers coming from other languages it can pose a challenge. If you are one of them, then you need to change your thinking a bit to understand asynchronous code.
+
+Note: It's still possible to write blocking code in Node. It won't prevent you from doing that and in some edge cases you want or have to write blocking code, e.g., cryptographic hashing of password. However in most cases non-blocking code should be used with one of the asynchronous patterns, e.g., callbacks or event emitters.
+
+Consider the following example, we have a few print statements in Ruby and with `sleep` to delay the output.
+
+```ruby
+puts "step 1"
+sleep(1)
+puts "step 2"
+```
+
+To run the file from this repository, execute `$ ruby sleep.rb`. You will see `"step 1"`, then the program will wait one second and you'll see `"step 2"`. So far so good. That's how synchronous code works because during the sleep method, nothing else is running. 
+
+On the contrary, in JavaScript and Node we can schedule some logic in the future (in a callback) and our program will be able to run something else while waiting for this future date/time. Look at this code with a `setTimeout()` which will output the same two lines as the Ruby code (second line with a 1 second delay):
+
+```js
+console.log('step 1')
+setTimeout(function(){
+  console.log('step 2')
+}, 1000)
+```
+
+However, if we had some code after asynchronous `setTimeout()`, the program will run that code first, before the callback delayed by 1 second. 
+
+```js
+console.log('step 1')
+setTimeout(function(){
+  console.log('step 2')
+}, 1000)
+console.log('step 1.5?')
+```
+
+The snippet above is in the `sleep.js` file. You can run it with `$ node sleep.js`. It will result in `"step 1"`, `"step 1.5"`, 1-second wait and then `"step 2"`. 
+
+Note: If you never heard a term callback, think about it as just a function. Often we define this function anonymously right in the other function's (like `setTimeout()`) call, meaning our callback doesn't have a name (example above), but nothing prevent us from declaring a named function or a variable which value is function and then using that name.
+
+The reason why I'm using `setTimeout()` is because it emulates time-intensive input/output operations where you don't have the results immediately and you still want your system to be functioning and doing something else while waiting (non-blocking I/O). For example, instead of `setTimeout()`, you would have `fs.readFile(path, callback)` to read from a file, or `db.users.find({}, callback)` to read from a database.
```

Comment:
what is `fs.readFile(path,callback)` - not familiar with fs... maybe remove code examples here and just say "like when you're reading from a file or pulling from the database"